### PR TITLE
propperly handle mineshaft changes 7294

### LINKED ISSUE #6537: Miner stop working after change mining level.
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.5
- MineColonies Version: minecolonies-0.13.610-ALPHA-universal
- Structurize Version: structurize-0.13.113-ALPHA-universal
- Related Mods and their Versions: no other mods.

## Expected behavior

Miner keep mining and find their own path in their current level setted. 

## Actual behavior

After playing for hours the miner stopped mining when I change the mining level. He went to another level (above or below) or even to another Hut and kept into a corner doing nothing.
I tried to reset the NPC, fire him and re-hiring him but he still stuck in this behaviour.
The problem was gone when I restart my Minecraft.

After restarting the game the miner get back to work.
image
The lost miner in another Hut.
image

## Steps to reproduce the issue

1. Get a miner lvl 4 (but another guy said with him this happened in lvl 3).
2. Keep playing for a long time (more than one or two hours).
3. Sometimes check where your miner are working. (do recall and then follow him)

### Logs

- latest.log:  (when this happen again I will upload the log)
- crashlog: 

### Notes

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #7294: Miner doesn't Mine on y:12 or y:8
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies for my Minecraft version.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.5 Forge: 36.1.30
- MineColonies Version: minecolonies-0.14.178-ALPHA-universal
Mine Colonies pack from Cursedforge with only addition of Smelting rotten flesh into Leather, had no problems so far

### Logs

- latest.log: https://gist.github.com/Ch4in3r007/c963cdfef8825364aac9311fc3a3bf1e

## Steps to reproduce the issue / details

So my Miner Level 3, Worker Level 8, housing Level 2 dug down his shaft and made the 2 last Nodes on y:12 and y:8, i assigned him to y:8, he dug up 5 Nodes, then stopped working, even after recall, Fire and hire etc.. So i asigned him to y:12 and he made 15 Nodes and also stopped working there. I led him dig a few Levels higher, there he had no problem. Today i tried again to set him down, he dug a few more nodes on y:8 and stopped again. after resetting him back to higher lvls and again back to the lowest lvl he went down, started digging (he got Cobbel in his Inventory), but after a few seconds stopped with request, as i opened he showed me error, i should open log and send it here. In the log the error is pretty much, but starting at Line 1300 it looks interesting.

Just reporting the AI issue. 

### Notes

---
## Viewers
grafik

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #7525: Dyer endless loop on already dyed armor pieces in inventory
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.5
- MineColonies Version: 0.14.341-ALPHA
- Structurize Version: 0.13.208-ALPHA
- Related Mods and their Versions: Forge 36.2.2

## Expected behavior
Dyer shall only try to dye virgin not yet dyed armour pieces.

## Actual behavior
Dyer is stuck in a loop moving back and forth between rack and hut worktable when a piece of already dyed armour is in the rack.

## Steps to reproduce the issue

1. Place order for three red dyed leather helmet
2. Dyer pick first virgin helmet and return to rack with it so dyed helmet is placed before remaining virgin helmets.
3. Dyer pick second virgin helmet, return with it and pick already dyed helmet
4. Dyer go to worktable and notice it is already dyed, so return to rack
5. Dyer now stuck unable to pick remaining virgin helmet until already dyed helmet is removed from rack.

### Notes

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #7529: Blacksmith doesn't "see" the warehouse stock
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.5
- MineColonies Version: 0.14.345-Alpha
- Structurize Version: 0.13.210-Alpha
- Related Mods and their Versions:

## Expected behavior

When the player requests a diamond axe via the post box, the blacksmith would request three diamonds and two sticks from the warehouse in order to craft said axe

## Actual behavior

When I requested the diamond axe, the request for diamonds ended up in the clipboard rather than as a delivery request.
The carpenter was asked to make the two sticks, while there are 1000 sticks in the warehouse.

I tried resetting the request system and restarting minecraft, and both didn't help.

Some screenshots: 

The warehouse stock:
Schermafbeelding 2021-08-24 om 00 20 59
Schermafbeelding 2021-08-24 om 00 21 29

Crafting recipe at the blacksmith:
Schermafbeelding 2021-08-24 om 00 24 06

Postbox gui:
Schermafbeelding 2021-08-24 om 00 26 17
Schermafbeelding 2021-08-24 om 00 26 44

Clipboard:
Schermafbeelding 2021-08-24 om 00 28 48

## Steps to reproduce the issue

1. Teach the blacksmith the diamond axe recipe
2. Make sure there are enough diamonds and sticks at the warehouse
3. Request a diamond axe at the post box
4. (In my case) watch it request the diamonds and sticks from anywhere but the warehouse
### Logs

The logs don't show request system requests.

- latest.log: 
- crashlog: 

### Notes

On discord, it looks like someone had a similar issue with a builder's hut. One of the builder's huts got all of the materials they needed, while the other one was requesting items to the player that are in the warehouse. Cancelling does requests didn't help.

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #7531: Healer (Hospital worker) gets stuck at their job block when hungry
Note: Playing on Valhesia 3 with Minecolonies on a server, so I can't upgrade minecolonies myself, so sorry if this is in the changelogs for the updates and I missed it, I gave a solid try to find it within the changelogs since that release.

### Prerequisites 

- [ ] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: Minecraft 1.16.5
- MineColonies Version: minecolonies-0.14.340-ALPHA
- Structurize Version: structurize-0.13.208-ALPHA
- Related Mods and their Versions: Modpack is Valhesia 3, version 3.4.2a

## Expected behavior
Healer, when in need of food, would leave the hospital, go to the restaurant, and await food there.
## Actual behavior
Healer sits at the job block (not where they recall to, but the other side of it), hungry and awaiting food. I'm using cave schematics for my town, and the hospital shares a wall with the restaurant.

## Steps to reproduce the issue

1. Create a cave hospital, level 5 (only know level 5 has the issue, can't be sure about below that)
2. place a cave restaurant directly next to the hospital, facing away
3. take all viable food from the healer, and wait until they get hungry.

### Logs

I wouldn't put much weight in this log, valhesia really spams errors a lot and i don't have a good way to filter down to minecolonies

- latest.log: https://gist.github.com/dmdinnc/82bd7153a43492fb7e88d3545cc3dcf5

### Notes
No additional notes.
---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #7294
Closes #6537
Closes #7525
Closes #7529
Closes #7531

# Changes proposed in this pull request:
- Miner should properly switch Mineshaft when they can't mine on the current anymore
- Miner should reset old target if the old target is not on the current level.
- Solve issues with tool like requests in schematics
- Solve issues with damagable requests
- Port sorting handling from Storageracks
- Not so shy at the restaurant

Review please

### CODE REVIEW

**Review by someaddons — DISMISSED** [review]

**Comment by firstiik** [comment]
minecolonies-0.14.345-ALPHA-universal.jar
The problem where the miner stood at the mine core and did not build the floor was solved.

**Review by someaddons — DISMISSED** [review]

**Comment by uecasm** [comment]
I haven't tested this, so take it with a grain of salt; but I wonder if the not-ignoring-damage for crafting requests, while good for things like armour crafting, might be problematic for sifter meshes and other tools with durability.

(There's already an outstanding reported issue relating to creating too many meshes because the sifter min stock doesn't recognise a damaged one as satisfying the stock, and the sifter doesn't prefer using the damaged mesh over a pristine replacement...)

**Comment by Raycoms** [comment]
> I haven't tested this, so take it with a grain of salt; but I wonder if the not-ignoring-damage for crafting requests, while good for things like armour crafting, might be problematic for sifter meshes and other tools with durability.
> 
> (There's already an outstanding reported issue relating to creating too many meshes because the sifter min stock doesn't recognise a damaged one as satisfying the stock, and the sifter doesn't prefer using the damaged mesh over a pristine replacement...)

I'll double check the sifter request for this.

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Stack.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ public Stack(@NotNull final ItemStack stack, final int count, final int minCount
      * Create a Stack deliverable.
      *
      * @param stack       the required stack.
-     * @param matchMeta   if meta has to be matched.
+     * @param matchMeta   if damage has to be matched.
```
Comment (uecasm) [inline src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Stack.java]: Rename the param too?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 2]
```diff
@@ -630,6 +630,11 @@ private IAIState searchANodeToMine(@NotNull final Level currentLevel)
         {
             workingNode = module.getActiveNode();
             module.setActiveNode(workingNode);
+
+            if (workingNode == null)
+            {
+                module.setCurrentLevel(module.getLevelId(currentLevel) + 1);
```
Comment (uecasm) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: Does this still work if they were already working the last level?  They do default to working the last level whenever they extend the shaft...

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Stack.java (hunk 1 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Stack.java]: adjusted

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: adjusted

**Review by uecasm — APPROVED** [review]

**Pull request merged**